lib/spack/spack/cmd/__init__.py

```diff
@@ -39,36 +38,39 @@ def python_name(cmd_name):
     return cmd_name.replace("-", "_")
 
 
+def require_python_name(pname):
+    """Require that the provided name is a valid python name"""
+    if python_name(pname) != pname:
+        raise PythonNameError(pname)
+
+
 def cmd_name(python_name):
     """Convert module name (with ``_``) to command name (with ``-``)."""
     return python_name.replace('_', '-')
 
 
-#: global, cached list of all commands -- access through all_commands()
-_all_commands = None
+def require_cmd_name(cname):
+    """Require that the provided name is a valid command name"""
```

Comment:
I can add the clarification. I wouldn't want to add any more checks here, since this was intended to be subservient to the `cmd_name()` function, which predates this PR. `require_cmd_name()` is literally asking the question, "Is the result of applying `cmd_name()` identical to that of `NOP`?" If we want `cmd_name()` to catch more unwanted characters than it does currently, I would suggest that should be the subject of a subsequent PR.